lib/spack/spack/database.py

```diff
@@ -419,14 +419,32 @@ class FailureTracker:
     the likelihood of collision very low with no cleanup required.
     """
 
+    #: root directory of the failure tracker
+    dir: pathlib.Path
+
+    #: whether the root directory has been created
+    created: bool
+
+    #: File for locking particular concrete spec hashes
+    locker: SpecLocker
+
     def __init__(self, root_dir: Union[str, pathlib.Path], default_timeout: Optional[float]):
         #: Ensure a persistent location for dealing with parallel installation
         #: failures (e.g., across near-concurrent processes).
         self.dir = pathlib.Path(root_dir) / _DB_DIRNAME / "failures"
-        self.dir.mkdir(parents=True, exist_ok=True)
+        self.created = False
 
         self.locker = SpecLocker(failures_lock_path(root_dir), default_timeout=default_timeout)
 
+    def _ensure_parent_directories(self) -> None:
+        """Ensure that parent directories of the FailureTracker exist.
+
+        Accesses the filesystem only once, the first time it's called on a given FailureTracker.
+        """
+        if not self.created:
+            self.dir.mkdir(parents=True, exist_ok=True)
```

Comment:
It's just one stat call to check if the dir exists if it was already created. Maybe that's worth it? Users can run `spack clean --failures` at any time.

Reply:
Ok see what you think now. Both the failure cache and the DB now do this... I worry about the DB but then again this is likely the least expensive thing the DB does.